src/integrations/cleanup-integration.php

```diff
@@ -25,63 +32,221 @@ public static function get_conditionals() {
+		$this->reset_cleanup();
+
+		$cleanups = $this->get_cleanup_tasks();
+		$limit    = $this->get_limit();
+
+		foreach ( $cleanups as $name => $action ) {
+			$items_cleaned = $action( $limit );
+
+			if ( $items_cleaned === false ) {
+				return;
+			}
+
+			if ( $items_cleaned < $limit ) {
+				continue;
+			}
+
+			// There are more items to delete for the current cleanup job, start a cronjob at the specified job.
+			$this->start_cron_job( $name );
+			return;
+		}
+	}
+
+	/**
+	 * Returns an array of cleanup tasks.
+	 *
+	 * @return Closure[] The cleanup tasks.
+	 */
+	protected function get_cleanup_tasks() {
+		return [
+			'clean_indexables_with_object_type_and_object_sub_type_shop-order' => function( $limit ) {
+				return $this->clean_indexables_with_object_type_and_object_sub_type( 'post', 'shop-order', $limit );
+			},
+			'clean_indexables_by_post_status_auto-draft' => function( $limit ) {
+				return $this->clean_indexables_with_post_status( 'auto-draft', $limit );
+			},
+			/* These should always be the last ones to be called. */
+			'clean_orphaned_content_indexable_hierarchy' => function( $limit ) {
+				return $this->cleanup_orphaned_from_table( 'Indexable_Hierarchy', 'indexable_id', $limit );
+			},
+			'clean_orphaned_content_seo_links_indexable_id' => function( $limit ) {
+				return $this->cleanup_orphaned_from_table( 'SEO_Links', 'indexable_id', $limit );
+			},
+			'clean_orphaned_content_seo_links_target_indexable_id' => function( $limit ) {
+				return $this->cleanup_orphaned_from_table( 'SEO_Links', 'target_indexable_id', $limit );
+			},
+		];
+	}
+
+	/**
+	 * Gets the deletion limit for cleanups.
+	 *
+	 * @return int The limit for the amount of entities to be cleaned.
+	 */
+	private function get_limit() {
+		/**
+		 * Filter: Adds the possibility to limit the number of items that are deleted from the database on cleanup.
+		 *
+		 * @api int $limit Maximum number of indexables to be cleaned up per query.
+		 */
+		$limit = \apply_filters( 'wpseo_cron_query_limit_size', 1000 );
 
-		if ( empty( $number_of_deletions ) ) {
-			$this->unschedule_cron( 'wpseo_cleanup_indexables' );
+		if ( ! \is_int( $limit ) ) {
+			$limit = 1000;
 		}
+
+		return $limit;
 	}
 
 	/**
-	 * Cleans orphaned rows from the yoast tables and unregisters the cron if no deletions.
+	 * Resets and stops the cleanup integration.
 	 *
 	 * @return void
 	 */
-	public function cleanup_orphaned_indexables() {
-		$deleted_orphans  = $this->cleanup_orphaned_from_table( 'Indexable_Hierarchy', 'indexable_id', 1000 );
-		$deleted_orphans += $this->cleanup_orphaned_from_table( 'SEO_Links', 'indexable_id', 1000 );
-		$deleted_orphans += $this->cleanup_orphaned_from_table( 'SEO_Links', 'target_indexable_id', 1000 );
+	public function reset_cleanup() {
+		\delete_option( self::CURRENT_TASK_OPTION );
+		\wp_unschedule_hook( self::CRON_HOOK );
+	}
 
-		$deleted_orphans = \apply_filters( 'wpseo_cleanup_orphaned', $deleted_orphans );
```

Comment:
@Xyfi This line used to enable Premium to delete orphaned entries from the Prominent Words table, whereas with the current PR, prominent words are not being taken into consideration, right?

Reply:
Fixed this by adding a new filter to be able to add new cleanup jobs to the cleanup integration.